google/cloud/storage/object_read_stream.cc

```diff
@@ -24,6 +25,17 @@ static_assert(std::is_move_assignable<ObjectReadStream>::value,
 static_assert(std::is_move_constructible<ObjectReadStream>::value,
               "storage::ObjectReadStream must be move constructible.");
 
+ObjectReadStream::ObjectReadStream()
+    : ObjectReadStream(absl::make_unique<internal::ObjectReadStreambuf>(
+          internal::ReadObjectRangeRequest("", ""),
+          Status(StatusCode::kUnimplemented, "null stream"))) {}
+
+ObjectReadStream::ObjectReadStream(ObjectReadStream&& rhs) noexcept
+    : std::basic_istream<char>(std::move(rhs)), buf_(std::move(rhs.buf_)) {
```

Comment:
I added some comments because I do not think there is a better way to express this, at least not in code (my wording could use improvement I bet).  We could try something like this:

```cc
ObjectReadStream::ObjectReadStream(ObjectReadStream&& rhs)
  : std::basic_istream<char>() {
  auto tmp = std::move(rhs.buf_);
  move(std::move(rhs));
  buf_ = std::move(tmp);
}
```

but the spec says:

> Effects: Move constructs from the rvalue rhs. This is accomplished by default constructing the base
> class, copying the gcount() from rhs, calling basic_ios<charT, traits>::move(rhs) to initialize
> the base class, and setting the gcount() for rhs to 0.

And there is no way to change `gcount()`.  Well, maybe with `swap()`:

```cc
ObjectReadStream::ObjectReadStream(ObjectReadStream&& rhs)
  : std::basic_istream<char>() {
  auto tmp = std::move(rhs.buf_);
  this->std::basic_istream<char>::swap(rhs);
  buf_ = std::move(tmp);
}
```

but that is even more obscure.  At the end of the day, the fact that `std::basic_istream<char>` is *designed* to support derived classes with move constructor is, I think, the strongest argument for just doing what I did.